readthedocs/proxito/views/mixins.py

```diff
@@ -30,52 +36,129 @@ class ServeDocsMixin:
 
     version_type = INTERNAL
 
-    def _serve_docs(
-            self,
-            request,
-            final_project,
-            path,
+    def _serve_docs(self, request, project, version, filename, check_if_exists=False):
+        """
+        Serve a documentation file.
+
+        :param check_if_exists: If `True` we check if the file exists before trying
+         to serve it.
+        """
+        base_storage_path = project.get_storage_path(
+            type_=MEDIA_TYPE_HTML,
+            version_slug=version.slug,
+            include_file=False,
+            version_type=version.type,
+        )
+
+        # Handle our backend storage not supporting directory indexes,
+        # so we need to append index.html when appropriate.
+        if not filename or filename.endswith("/"):
+            filename += "index.html"
+
+        # If the filename starts with `/`, the join will fail,
+        # so we strip it before joining it.
+        storage_path = build_media_storage.join(base_storage_path, filename.lstrip("/"))
+
+        if check_if_exists and not build_media_storage.exists(storage_path):
+            return None
+
+        self._track_pageview(
+            project=project,
+            path=filename,
+            request=request,
             download=False,
-            version_slug=None,
-    ):
+        )
+        return self._serve_file(
+            request=request,
+            storage_path=storage_path,
+            storage_backend=build_media_storage,
+        )
+
+    def _serve_dowload(self, request, project, version, type_):
         """
-        Serve documentation in the way specified by settings.
+        Serve downloadable content for the given version.
+
+        The HTTP header ``Content-Disposition`` is added with the proper
+        filename (e.g. "pip-pypa-io-en-latest.pdf" or "pip-pypi-io-en-v2.0.pdf"
+        or "docs-celeryproject-org-kombu-en-stable.pdf").
+        """
+        storage_path = project.get_storage_path(
+            type_=type_,
+            version_slug=version.slug,
+            version_type=version.type,
+            include_file=True,
+        )
+        self._track_pageview(
```

Comment:
Is there any reason why we do the `_track_pageview` here but the `analytics_event.delay` in the outer function at https://github.com/readthedocs/readthedocs.org/pull/9726/files#diff-21e4e802b7b57cbdcc4fd900b2ab89fa6c7b43333ae47f7ad14b489a3bed66b8R367? Why not both at the same place?

Reply:
O even maybe just on `_serve_file` directly?

Reply:
I think it's better to call this function only once from `_serve_file` that remember to call it from all the other methods. It seems the only difference here is the `download=True` argument, that we can pass to `_serve_file` to only make one call to this method.

Reply:
> Why not both at the same place?

Just old code

> I think it's better to call this function only once from _serve_file

This should be called only when downloading files, `_serve_file` does the serving of any file, even static files.